### PR TITLE
Add map state in the Go Sdk

### PR DESCRIPTION
**Pull request by damccorm**
This continues to grow our state support by adding map state support. Once this is in, I'll add a follow up PR with an integration test. In the meantime, I tested manually by running the hacked wordcount found here - https://github.com/damccorm/beam/pull/94/files#diff-46f0bd8f9f6b541e840079960d684ba78e6513f1293b673d08faaec94d48a234 - and it produced output like:

```
any, 1. Total num keys: 2. With keys: [any5 any]
any, 2. Total num keys: 3. With keys: [any7 any5 any]
any, 3. Total num keys: 4. With keys: [any28 any7 any5 any]
any, 4. Total num keys: 5. With keys: [any28 any7 any5 any any13]
any, 5. Total num keys: 6. First 5 keys: [any28 any7 any5 any any13]
any, 6. Total num keys: 7. First 5 keys: [any28 any7 any5 any any13]
any, 7. Total num keys: 8. First 5 keys: [any28 any5 any13 any97 any98]
any, 8. Total num keys: 9. First 5 keys: [any28 any5 any13 any97 any98]
any, 9. Total num keys: 10. First 5 keys: [any28 any5 any13 any97 any98]
any, 10. Total num keys: 11. First 5 keys: [any28 any5 any13 any97 any98]
any, 11. Total num keys: 12. First 5 keys: [any28 any5 any13 any97 any98]
any, 12. Total num keys: 13. First 5 keys: [any28 any5 any13 any97 any98]
any, 13. Total num keys: 14. First 5 keys: [any28 any5 any13 any97 any98]
any, 14. Total num keys: 14. First 5 keys: [any28 any5 any13 any97 any98]
any, 15. Total num keys: 15. First 5 keys: [any33 any8 any28 any5 any6]
any, 16. Total num keys: 16. First 5 keys: [any33 any8 any28 any5 any6]
any, 17. Total num keys: 17. First 5 keys: [any33 any9 any8 any28 any5]
any, 18. Total num keys: 18. First 5 keys: [any33 any9 any8 any28 any5]
Camelot, 1. Total num keys: 2. With keys: [Camelot Camelot17]
Mean, 1. Total num keys: 2. With keys: [Mean87 Mean]
Reverbs, 1. Total num keys: 2. With keys: [Reverbs Reverbs73]
devour, 1. Total num keys: 2. With keys: [devour devour26]
vengeance, 1. Total num keys: 2. With keys: [vengeance vengeance6]
vengeance, 2. Total num keys: 3. With keys: [vengeance72 vengeance vengeance6]
vengeance, 3. Total num keys: 4. With keys: [vengeance64 vengeance6 vengeance72 vengeance]
...
```

Note that this doesn't add support for Remove (to remove a single key). This will come in a future pr and has been noted in #22736

Part of #22736

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @riteshghorse 

FWIW, this is a larger PR because I thought it would make more sense as a cohesive change and because its so similar to the other changes. With that said, as you review I'd probably recommend thinking about it in 3 pieces (and reviewing it in that order):

1. Extending existing plumbing to support keyed state (both translate.go's)
2. Execution side stuff (statemgr.go + userstate.go)
3. User side stuff (state.go + fn.go).
4. Misc interface + tests (everything else)

I'd probably recommend reading the PR in that order.

Generally, the reason this PR is a little bigger is because we're using new protos for multimap state instead of the ones we've been using for bag state.

File: sdks/go/pkg/beam/core/runtime/exec/userstate.go [hunk 1]
```diff
@@ -106,12 +110,12 @@ func (s *stateProvider) WriteValueState(val state.Transaction) error {
 	return nil
 }
 
-// ReadBagState reads a ReadBagState state from the State API
+// ReadBagState reads a BagState state from the State API
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/runtime/exec/userstate.go:113]:
```suggestion
// ReadBagState reads a bag state from the State API
```

File: sdks/go/pkg/beam/core/runtime/exec/userstate.go [hunk 2]
```diff
@@ -162,6 +165,105 @@ func (s *stateProvider) WriteBagState(val state.Transaction) error {
 	return nil
 }
 
+// ReadMapStateValue reads a value from the map state given a key.
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/runtime/exec/userstate.go:168]:
```suggestion
// ReadMapStateValue reads a value from the map state for a given map key.
```

File: sdks/go/pkg/beam/core/state/state_test.go [hunk 3]
```diff
@@ -116,6 +117,50 @@ func (s *fakeProvider) ExtractOutputFn(userStateID string) reflectx.Func {
 	return nil
 }
 
+// ReadMapStateValue(userStateID string, key interface{}) (interface{}, []Transaction, error)
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/state/state_test.go:120]: Did you mean to add doc comments for the functions below?

Comment (damccorm) [reply sdks/go/pkg/beam/core/state/state_test.go]: Whoops, no this is just a leftover bad copy paste as I moved the definitions over. I shouldn't need doc comments for test functions, but I did remove these. Good catch!

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go [hunk 4]
```diff
@@ -1376,3 +1398,13 @@ func UpdateDefaultEnvWorkerType(typeUrn string, pyld []byte, p *pipepb.Pipeline)
 	}
 	return errors.Errorf("unable to find dependency with %q role in environment with ID %q,", URNArtifactGoWorkerRole, defaultEnvId)
 }
+
+// UserStateCoderId returns the coder id of a user state
+func UserStateCoderId(ps state.PipelineState) string {
+	return fmt.Sprintf("val_%v", ps.StateKey())
+}
+
+// UserStateCoderId returns the key coder id of a user state
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/runtime/graphx/translate.go:1407]:
```suggestion
// UserStateKeyCoderId returns the key coder id of a user state
```

**Review by riteshghorse — APPROVED** [review]
Thanks, this looks great. One comment fix(missed this one before).

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go (hunk 4 above)
Comment (damccorm) [reply sdks/go/pkg/beam/core/runtime/graphx/translate.go]: Thanks! Hopefully that saves a follow up linting PR 😅

**Pull request merged**